broker/lib/fabrik/DirectorManager.js

```diff
@@ -177,7 +188,205 @@ class DirectorManager extends BaseManager {
+        throw new DeploymentDelayed(deploymentName);
+      }
+    } else {
+      // user-triggered operations
+      if (shouldRunNow) {
+        //check if the deployment already exists in etcd (poller-run)
+        return boshOperationQueue.containsDeployment(deploymentName).then(enqueued => {
+          if (enqueued) {
+            results.cached = true;
+          }
+          return results;
+        });
+      } else {
+        results.enqueue = true;
+        return results;
+      }
+    }
+  }
+
   createOrUpdateDeployment(deploymentName, params, args) {
+    logger.info(`Checking rate limits against bosh for deployment `);
+    const previousValues = _.get(params, 'previous_values');
+    const action = _.isPlainObject(previousValues) ? CONST.OPERATION_TYPE.UPDATE : CONST.OPERATION_TYPE.CREATE;
+    const scheduled = _.get(params, 'scheduled') || false;
+    const runImmediately = _.get(params, '_runImmediately') || false;
+    _.omit(params, 'scheduled');
+    _.omit(params, '_runImmediately');
+
+    if (!config.enable_bosh_rate_limit || runImmediately) {
+      return this._createOrUpdateDeployment(deploymentName, params, args, scheduled)
+        .then(taskId => {
+          return {
+            cached: false,
+            task_id: taskId
+          };
+        });
+    }
+    const decisionMaker = {
+      'shouldRunNow': false,
+      'cached': false
+    };
+    return this.executePolicy(scheduled, action, deploymentName)
+      .then(checkResults => this.enqueueOrTrigger(checkResults.shouldRunNow, scheduled, deploymentName))
+      .tap(res => {
+        decisionMaker.shouldRunNow = res.shouldRunNow;
+        decisionMaker.cached = res.cached;
+      })
+      .then(res => res.enqueue)
+      .then(enqueue => {
+        if (enqueue) {
+          // stagger here by putting it into etcd cache and return promise
+          return boshOperationQueue.saveDeployment(this.plan.id, deploymentName, params, args)
+            .then(() => {
+              decisionMaker.cached = true;
+              throw new DeploymentDelayed(deploymentName);
+            });
+        }
+      })
+      .then(() => {
+        if (decisionMaker.shouldRunNow && decisionMaker.cached) {
+          //the deployment was cached in etcd earlier; remove it from cache and proceed
+          return boshOperationQueue.deleteDeploymentFromCache(deploymentName);
+        }
+      })
+      .then(() => this._createOrUpdateDeployment(deploymentName, params, args, scheduled))
+      .then(taskId => {
+        const out = {
+          'cached': false
+        };
+        if (decisionMaker.cached) {
+          return boshOperationQueue.saveBoshTask(this.getInstanceGuid(deploymentName), taskId)
+            .then(() => {
+              out.cached = true;
+              out.task_id = taskId;
+              return out;
+            });
+        } else {
+          out.task_id = taskId;
+          return out;
+        }
+      })
+      .catch(DeploymentDelayed, e => {
```

Comment:
could have avoided this promise chain by doing something like this:

```
 if(enque){
   // Do what is required & return 'cached' true
}else {
 return Promise.try(()=> {
   if(shouldRunNow & Cached) {
      //delete cache entry
    }
    }).then(() => this._createOrUpdateDeployment(deploymentName, params, args, scheduled))
       ...
}

```
keeps the flow in a logical order. had this exception `DeploymentDelayed` been thrown from elsewhere as it would have made sense to have it this way, but this is similar to a `goto` stmt which could be avoided.

Reply:
That's true but this leads to inner promise chains being elongated. Of course, I am not questioning this because the way to break out of promise chains early is a long-running discussion (https://github.com/petkaantonov/bluebird/issues/581). In fact, the creator of Bluebird prefers throwing an exception to break out [go figure! :) ]
I would keep this flow for now and refactor this in a latter PR. Hope this is fine

Reply:
ok